WalletWasabi.Daemon/WasabiAppBuilder.cs

```diff
@@ -0,0 +1,187 @@
+	}
+
+	private Global CreateGlobal()
+	{
+		var walletManager = new WalletManager(Config.Network, Config.DataDir, new WalletDirectories(Config.Network, Config.DataDir));
+		return new Global(Config.DataDir, Config, walletManager);
+	}
+
+	private PersistentConfig LoadOrCreateConfigs()
+	{
+		Directory.CreateDirectory(Config.DataDir);
+
+		PersistentConfig persistentConfig = new(Path.Combine(Config.DataDir, "Config.json"));
+		persistentConfig.LoadFile(createIfMissing: true);
+
+		if (persistentConfig.MigrateOldDefaultBackendUris())
+		{
+			// Logger.LogInfo("Configuration file with the new coordinator API URIs was saved.");
+			persistentConfig.ToFile();
+		}
+
+		return persistentConfig;
+	}
+
+	private void TaskScheduler_UnobservedTaskException(object? sender, UnobservedTaskExceptionEventArgs e)
+	{
+		AppConfig.UnobservedTaskExceptionsEventHandler?.Invoke(this, e.Exception);
+	}
+
+	private void CurrentDomain_UnhandledException(object? sender, UnhandledExceptionEventArgs e)
+	{
+		if (e.ExceptionObject is Exception ex)
+		{
+			AppConfig.UnhandledExceptionEventHandler?.Invoke(this, ex);
+		}
+	}
+
+	private async Task TerminateApplicationAsync()
+	{
+		Logger.LogSoftwareStopped(AppConfig.AppName);
+
+		if (Global is { } global)
+		{
+			await global.DisposeAsync().ConfigureAwait(false);
+		}
+	}
+
+	private void SetupLogger()
+	{
+		LogLevel logLevel = Enum.TryParse(Config.LogLevel, ignoreCase: true, out LogLevel parsedLevel)
+			? parsedLevel
+			: LogLevel.Info;
+		Logger.InitializeDefaults(Path.Combine(Config.DataDir, "Logs.txt"), logLevel);
+	}
+}
+
+public record WasabiAppBuilder(string AppName, string[] Arguments)
+{
+	internal bool MustCheckSingleInstance { get; init; }
+	internal EventHandler<Exception>? UnhandledExceptionEventHandler { get; init; }
+	internal EventHandler<AggregateException>? UnobservedTaskExceptionsEventHandler { get; init; }
+	internal Action Terminate { get; init; } = () => { };
+
+	public WasabiAppBuilder ApplicationName(string appName) =>
+		this with { AppName = appName };
+
+	public WasabiAppBuilder EnsureSingleInstance(bool ensure = true) =>
+		this with { MustCheckSingleInstance = ensure };
+
+	public WasabiAppBuilder OnUnhandledExceptions(EventHandler<Exception> handler) =>
+		this with { UnhandledExceptionEventHandler = handler };
+
+	public WasabiAppBuilder OnUnobservedTaskExceptions(EventHandler<AggregateException> handler) =>
+		this with { UnobservedTaskExceptionsEventHandler = handler };
+
+	public WasabiAppBuilder OnTermination(Action action) =>
+		this with { Terminate = action };
+	public WasabiApplication Build() =>
+		new(this);
+
+	public static WasabiAppBuilder Create(string appName, string[] args) =>
+		new(appName, args);
+}
+
+public static class WasabiAppExtensions
+{
+	public static async Task<ExitCode> RunAsConsoleAsync(this WasabiApplication app)
+	{
+		return await app.RunAsync(
+			async () =>
+			{
+				await app.Global!.InitializeNoWalletAsync(app.TerminateService);
+
```

Comment:
hahaha absolutely yes!

Reply:
Would be great to notify systemd too.

Reply:
@Kixunil what do you mean exactly?

Reply:
There are probably libraries to just call a function when the daemon is ready (most popular langs have them, probably including C#) but what the function does is checking if there's a specific environment variable and if there is just open a unix socket at path from that variable and write something specific to it. Pretty simple stuff.

This can help coordinate multiple daemons during system boot etc.

Reply:
I see. We still dont know how other will use it. I expect to discuss this kind of things with some Wasabi friends that are interested in this feature to try to understand better how they use it or how they would like to use it. Right now they use containers to isolate the different instances, something that shouldn't be really necessary. I will ask them this.